Playwright: Move components and objects off BaseContainer.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR proposes to implement Solution A as posted in [this comment](https://github.com/Automattic/wp-calypso/pull/54697#issue-692864875) for Components and Pages in the`@automattic/calypso-e2e` module.

Key changes:
- removal of inheritance on BaseContainer and associated overridden methods such as `Expect` and `_postInit`.

#### Testing instructions

- [x] Playwright mobile tests continue to run and pass.
- [x] Playwright desktop tests continue to run and pass.
- [x] Unit tests continue to pass.

Related to #54688

worldomonation committed:
Move components off BaseContainer.
- navbar component
- notifications component

worldomonation committed:
Remove references to BaseContainer for all components.

worldomonation committed:
Update tests.

worldomonation committed:
Remove Expect from another test

worldomonation committed:
Move PublishedPosts pages and Login page off BaseContainer

worldomonation committed:
Move rest of the POMs off BaseContainer

worldomonation committed:
Migrate GutenbergEditorPage off BaseContainer.

worldomonation committed:
Refactor GutenbergEditorPage

worldomonation committed:
Fix test case

worldomonation committed:
Remove BaseContainer file.

dpasque commented on packages/calypso-e2e/src/lib/components/comments-component.ts:
```diff
@@ -15,10 +14,28 @@ const selectors = {
+	 *
+	 * @returns {Promise<void>} No return value.
+	 */
+	async waitUntilLoaded(): Promise< void > {
+		await this.page.waitForLoadState( 'load' );
+	}
+
```
Slight nit: for this POM, and a few similar other ones in these changes, do you think we need this method if all we're doing is waiting for the "load" state on the overall page?

This method title gives the impression to me that this is waiting for something specific to the _component itself_ that we're waiting on, but really we're just waiting on the overall page to settle. 

I like the pattern you've set up where some components expose the same `waitUntilLoaded` - it's really nice consistency! But my gut is that we should only use that if we're actually waiting for something specific to the component / page at hand. I just wouldn't want folks to have any wrong assumptions calling this! 

Comparatively, for when we still need this wait, a `page.waitForLoadState("load")` in the script itself right after a navigation might make it really clear that we need the page to settle before continuing on. Or maybe for components that will often need to be extra careful and hedge against this, we could use a private method to that class?

worldomonation replied:
> Comparatively, for when we still need this wait, a page.waitForLoadState("load") in the script itself right after a navigation might make it really clear that we need the page to settle before continuing on.

To make sure I am understanding correctly, is the suggestion to have the test script writer call for a `wait` in the script?

> Or maybe for components that will often need to be extra careful and hedge against this, we could use a private method to that class?

I think this would be much more preferable. Stylistically I would prefer that test scripts contain as little of the inner workings of the framework as possible (though there are obviously exceptions existing in the codebase). This predominantly the reason why I have decided to include the `waitForLoadState` for any methods that may need to wait prior to proceeding. 

A `private` method that is called by any class method needing to ensure a proper state sounds like a good compromise. To make sure we are on the same page, this is what I am imagining:

```
private async waitForLoadState() {
    await page.waitForLoadState( 'whatever' );
}


async methodThatNeedsToSettle() {
    await this.waitForLoadState();
    await blah
}

```

This would leave the public facing method that could be named like `waitUntilLoaded` which would handle waiting for and returning an element, or performing any additional tasks, etc.

dpasque replied:
Yes that would definitely work!

The short of what I was saying was that I wasn't sure we should have a public method on a POM class that _only_ waited for general `page` navigation specifically (which seems to me to be in a somewhat different category than waiting on specific elements and their stability). 

I think it makes more sense to place those waits more privately in classes where needed, and make them clear that they are related to `page` loading, just as you implemented! I think you're new changes look really great! :)

worldomonation replied:
@dpasque I've made some changes, take a look:

1. remove `waitUntilLoaded` wherever possible.

Turns out a lot of the places where I inserted that, it was not actually necessary since the subsequent calls eg. `click` would have handled it already.

2. rename `waitUntilLoaded` to `pageSettled`.

I think this would reflect what we are trying to do with this method better.

3. make `pageSettled` a private method.

This way test script writer will not accidentally call this method expecting any other side effects.

dpasque replied:
Yep, just saw the changes, and I think it reads so well! Thanks for making the updates!

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -33,23 +32,39 @@ const selectors = {
+	 */
+	async waitUntilLoaded(): Promise< Frame > {
+		const frame = await this.getEditorFrame();
 		await this.page.waitForLoadState( 'networkidle' );
```
Should we put this call above `getEditorFrame()`? Is it worth letting things fully settle before trying to do any kind of selecting?

worldomonation replied:
Ah yes, good find!

dpasque reviewed (changes requested):
Hurray!! These changes look so good, thank you so much for making these updates! 🎉 

I have just two pattern questions - otherwise LGTM!

worldomonation committed:
Code review comments:
- remove `waitUntilLoaded` wherever possible.
- rename `waitUntilLoaded` to `pageSettled` and make the method private.

dpasque reviewed (approved):
Looks perfect!! Thanks again for making all these updates! 🎉

worldomonation commented:
Received approval but waiting for all tasks to pass before merging.

worldomonation committed:
Restore LoginPage's original load state
